File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.py
'''
1. 아이디어
    - 한 장씩 뽑고 다시 사용할 수 없는 걸 보아 queue 활용 가능성
    - goal에서 하나씩 가져와서, cards1, cards2에 있으면 제거
        - goal이 없어지기 전에 queue가 먼저 없어지거나, 꺼낸 것과 다르다면 "NO"
2. 변수
- queue: cards1, cards2, goal
'''

from collections import deque

def solution(cards1, cards2, goal):
    
    cards1 = deque(cards1)
    cards2 = deque(cards2)
    goal = deque(goal)
    
    while goal:
        
        item = goal.popleft()
        if cards1 and item == cards1[0]:
            cards1.popleft()
        elif cards2 and item == cards2[0]:
            cards2.popleft()
        else:
            return "No"
    return "Yes"
    

File: 프로그래머스/1/159994. 카드 뭉치/카드 뭉치.java
import java.util.*;

class Solution {
    public String solution(String[] cards1, String[] cards2, String[] goal) {
        
        Queue<String> q1 = new LinkedList<> (Arrays.asList(cards1));
        Queue<String> q2 = new LinkedList<> (Arrays.asList(cards2));
        
        for(String str : goal){
            if(!q1.isEmpty() && q1.peek().equals(str))
                q1.poll();
            else if(!q2.isEmpty() && q2.peek().equals(str))
                q2.poll();
            else
                return "No";
        }
        
        return "Yes";
    }
}